### ISSUE TITLE
Can't build for iOS with Xamarin Forms[BUG] 

### ISSUE THREAD

**Issue by Pequeno0** [mono/SkiaSharp#1702]
**Description**

Including SkiaSharp in the Xamarin forms project, Windows paired to a Mac, then trying to build yields the following error:

1>C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\Xamarin\iOS\Xamarin.Shared.targets(975,3): error : DirectoryNotFoundException: Could not find a part of the path '/Users/<username>/Library/Caches/Xamarin/mtbs/builds/TestiOSBuildApp.iOS/8c512dc4f4826b88a51c8798b7f65927dc2afe3d8cd845a563df3fddd5f3932e/bin/iPhone/Debug/TestiOSBuildApp.iOS.app/Frameworks/libSkiaSharp.framework/_CodeSignature'.

which means it won't build the app. 

It seems this _CodeSignature is not copied from the nuget package for some reason?

**Code**

Make a new Xamarin Forms project in Visual Studio, get the SkiaSharp package from Nuget, pair to a mac from your PC, build, see the the failure.

**Expected Behavior**

The _CodeSignature folder should exist.

**Actual Behavior**

_CodeSignature folder is missing, and the the build fails.

**Basic Information**

- Version with issue: 2.80.2
- Last known good version:  2.80.2
- IDE:  Microsoft Visual Studio Enterprise 2019, Version 16.9.5
- Platform Target Frameworks: Xamarin Forms Portable build to iOS.
  - iOS: 11
- Target Devices:   iPhone 7, Software version 14.4.2

**Comment by taublast** [comment]
Been fighting this few days, thought that was coming from my settings. But after reformatting reinstalling Mac from scratch and having set up all the certificates again, it's the same: it's compiling-signing ok for real device until i add SkiaSharp nuget.
This appeared after upgrading to XCode 12.5. Still don't know if this is a new apple's requirement for third-party libs to be imperatively signed OR a visual studio build bug OR xcode bug.. Any thoughts?
This is a very urgent matter..

**Comment by Grzesik** [comment]
Hi, I have the same problem! I have two projects with SkiaSharp on two comupters and I get the same error in iOS project. Android works. I noticed also, that the common Project shows following warning:
warning NU1701: Package 'SkiaSharp.Views 2.80.2' was restored using '.NETFramework,Version=v4.6.1, .NETFramework,Version=v4.6.2, .NETFramework,Version=v4.7, .NETFramework,Version=v4.7.1, .NETFramework,Version=v4.7.2, .NETFramework,Version=v4.8' instead of the project target framework '.NETStandard,Version=v2.0'. This package may not be fully compatible with your project.

I deletes all Nuget packages and reinstalled them in the project. Nothing helps.

Thanks

**Comment by mrutter** [comment]
Hello everybody.

I have exacty the same issue here, trying to run the iOS simulator from Visual Studio (Windows) paired with the Mac.
The weird thing is that if I compile/debug from Visual Studio for Mac, everything works fine.

Right now I'm not able to debug from Visual Studio nad I have to work on Mac :(

Thank you.
-marcello

**Comment by monster74** [comment]
Hello.
I have the same problem with SkiaSharp after updating xcode 12.5 and VS 16.9.5.
`DirectoryNotFoundException: Could not find a part of the path '/Users/xxxxxx/Library/Caches/Xamarin/mtbs/builds/App.iOS/6635a42b07b773f91fc2b39356bd40a5a6aa09cd70010ccd9d3e89de3a4c1a2a/bin/iPhone/Debug/device-builds/iphone9.3-14.5.1/App.iOS.app/Frameworks/libSkiaSharp.framework/_CodeSignature'.`
Projects without SkiaSharp build normally.

**Comment by taublast** [comment]
Additional info after installing VS 16.10.0 preview 4 + Xamarin.iOS '14.19.0.7'

1. VS 16.9.6 + Xamarin.iOS '14.19.0.7 = same issue as above

2. VS 16.10.0 preview 4 + Xamarin.iOS '14.19.0.7' = 

Compiling signing finishes without errors, when trying to deploy to:

a - simulator: fails without any error description
b - real device: fails with:

>error MT1006: Could not install the application '/Users/Airboom/Library/Caches/Xamarin/mtbs/builds/XamarinTestApp.iOS/31c782e90715fd993ed01e15963fc6f699e2dc9661442e7b92edfc01c6ad08cb/bin/iPhone/Debug/XamarinTestApp.iOS.app' on the device 'iPhone nicky': Package Inspection Failed Error (error: 0xe8000051).	 

> PackageInspectionFailed: Failed to load Info.plist from bundle at path /private/var/installd/Library/Caches/com.apple.mobile.installd.staging/temp.08ASXq/extracted/XamarinTestApp.iOS.app/Frameworks/libSkiaSharp.framework``

Well to me it looks like a Visual Studio for Windows bug. Agree?

**Comment by peder1896** [comment]
Hi
I have the same problem I can deploy from Visual Studio for Mac 8.9.10 (build 4)
but not from Visual Studio 16.9.6 I then get the foolowing error:
 MessagingRemoteException: An error occurred on client Build169000294 while executing a reply for topic xvs/build/execute-task/xxxxxx.iOS/05a14ae002fCodesign
AggregateException: One or more errors occurred.
DirectoryNotFoundException: Could not find a part of the path '/Users/xxxxxxx/Library/Caches/Xamarin/mtbs/builds/xxxxxx.iOS/05a14ae99af88b6e851484e35f30dfcb0d88549771d3e6791c11d2d825147791/bin/iPhone/Debug/xxxxxxx.iOS.app/Frameworks/libSkiaSharp.framework/_CodeSignature'

**Comment by cbs-cbt** [comment]
Hello,
Same problem here.
Tried with a brand new Xamarin.Forms solution.
As soon as SkiaSharp.Views.Forms NuGet package is added to the solution iOS project won't compile.
But downgrading to SkiaSharp.Views.Forms version 1.68.3 will work again

**Comment by gureAngel** [comment]
> Hello,
> Same problem here.
> Tried with a brand new Xamarin.Forms solution.
> As soon as SkiaSharp.Views.Forms NuGet package is added to the solution iOS project won't compile.
> But downgrading to SkiaSharp.Views.Forms version 1.68.3 will work again


I tried downgrading the version to 1.68.3 and it still doesn't work.

In the emulator does not give me any error, but when I want to run it on the physical device I still have the same error:

MessagingRemoteException: An error occurred on client Build169000294 while executing a reply for topic xvs/build/execute-task/xxxxxxxxxxxx.iOS/ed22858002fCodesign
AggregateException: One or more errors occurred.
DirectoryNotFoundException: Could not find a part of the path '/Users/xxxx/Library/Caches/Xamarin/mtbs/builds/xxxxxxx.iOS/xxxxxxxxxx/bin/iPhone/Debug/xxxxxxxx.iOS.app/Frameworks/libSkiaSharp.framework/_CodeSignature'.

**Comment by dmsi-leo** [comment]
Hi there,

Having the same issue with SkiaSharp.Views.Forms 2.80.2.
iOS build failed with the error:
DirectoryNotFoundException: Could not find a part of the path '/Users/developer/Library/Caches/Xamarin/mtbs/builds/MInspect.iOS/3f461d8a077ea1533355063303276a96508c09fa0c4d9422a2d65b43a569ef81/bin/iPhone/Debug/MInspect.iOS.app/Frameworks/libSkiaSharp.framework/_CodeSignature'.

VS 2019 16.9.6
XF v5.0.0.2012

Works fine if I rolled back SkiaSharp.Views.Forms to v1.68.3.

Thanks

**Comment by taublast** [comment]
Can confirm it compiles, signs and deploys OK on real iPhone with an older SkiaSharp.Views.Forms v1.68.3 🙉

Tested with VS 16.9.6 + Xamarin.iOS '14.19.0.7 + XCode 12.5

**Comment by gureAngel** [comment]
Downloading the project, reloading it, downgrading the version to 1.68.3, it worked again.

**Comment by IGx89** [comment]
In case it helps, I ran into the exact same issue with https://www.nuget.org/packages/Xamarin.Google.iOS.Cast/. Strange thing is, I was on 16.9.6 and it was working, but after upgrading to 16.10.0 and downgrading back to 16.9.6 (due to another Xamarin bug) this issue started happening. Could have been triggered by me clearing cache/build folders as part of the upgrade/downgrade.

**Comment by ahoke-cr** [comment]
I had this issue pop up randomly today when trying to deploy to my iPhone instead of the Simulator. Now I cannot fix it.

**Comment by taublast** [comment]
After upgrading Visual Studio to 16.10.0 + Xamarin.iOS 14.20.0.1 it finally builds and signs ok, from Windows, using SkiaSharp.Views.Forms 2.80.2.

Still not sure was it VS ok Xamarin.iOS that caused the issue.
Had another problem with build that appeared with this VS update but that's a different story, guess we can clause this one.

**Comment by joehanna** [comment]
I have the same problem on VS Mac (latest version) but I don't even use SkiaSharp.  Release mode is fine on both simulator and device, but not Debug. Wondering if it is not related to SkiaSharp at all.

**Comment by ahoke-cr** [comment]
> I have the same problem on VS Mac (latest version) but I don't even use SkiaSharp. Release mode is fine on both simulator and device, but not Debug. Wondering if it is not related to SkiaSharp at all.

If you delete the build folder for release I think you will encounter the issue on release as well. It only started happening after I cleared my build folder when trying to fix another issue.

**Comment by boris-df** [comment]
Same Issue:
- VS 16.10
- Xcode 12.5
- VS for Mac 8.10 Preview (build 1772) … also problems with the stable version, so I switched to preview to see if helps.
- Mono 6.12.0.140
- Xamarin.iOS: 14.20.0.1 

Just create a new App-Project (for example with the tabbed app template)
- it compiles and runs in the simulator

Now add SkiaSharp Package Version 2.x upwards
- depending on the exact package version you get errors like: "Package inspection failed  (error: 0xe8000051)" or with the newer packages some weird errors in the device-log  (not sure though if they are all related to skia) like:
"Invalid value for UNNotificationExtensionCategory in Info.plist"
"Invalid value for UNNotificationExtensionCustomSectionIdentifier in Info.plist"
"Unable to initialize CXCallDirectoryStore for reading and writing: Error Domain=NSCocoaErrorDomain Code=513 "You don’t have permission to save the file “CallDirectory” in the folder “Library”." UserInfo={NSFilePath=/var/mobile/Library/CallDirectory, NSUnderlyingError=0x13be04820 {Error Domain=NSPOSIXErrorDomain Code=13 "Permission denied"}}"
"Failed to set fraction completed: 0.00 for: com.companyname.App5"
…

When you change the SkiaSharp Package to Version 1.68.3 - Die app compiles and starts again.


FWIW - VS 16.10 seems to have it's own bugs too! So I see that the apps doesn't have icons anymore and someone wrote, that the cannot upload Builds to Apple anymore because of this (see: https://developercommunity.visualstudio.com/t/Cant-deliver-to-the-appstore-since-last/1436223)

I don't know if this has to do with the problems we see with skiasharp 2.x currently

**Comment by ysmoradi** [comment]
16.11 Preview 1
Xamarin iOS 14.20.0.1
XCode 12.5 (12E262)

The same issue )":

**Comment by Elinareso** [comment]
Same issue after upgrade x-Code to 12.5 and VS 16.10, Skia 2.80.0

Severity	Code	Description	Project	File	Line	Suppression State
Error		MessagingRemoteException: An error occurred on client Build169000292 while executing a reply for topic xvs/build/execute-task/ProjectName/c8561a8002fCodesign
AggregateException: One or more errors occurred.
DirectoryNotFoundException: Could not find a part of the path '../c8561a8a7bfbc1a65b4c3ab6271481afb331465d1040423802a41c646a20d0bc/bin/iPhone/Debug/device-builds/iphone10.2-14.4.2/../Frameworks/libSkiaSharp.framework/_CodeSignature'.

**Comment by ysmoradi** [comment]
> 16.11 Preview 1
> Xamarin iOS 14.20.0.1
> XCode 12.5 (12E262)
> 
> The same issue )":

Note that on the same macos machine that I connect to it while developing iOS apps, I've installed the latest stable version of vs for mac, and vs for mac is able to build / run the exact same project successfully.

**Comment by MDGMDGMDG** [comment]
Uninstalling nuget packages v. 2.80.2 of SkiaSharp, SkiaSharp.Views and SkiaSharp.Views.Forms and reinstalling v. 1.68.3 I was able to build run debug on Simulator (only directly on the Mac server).

Visual Studio 2019 16.10
Xcode 12.5
Xamarin iOS 14.20.0.1

**Comment by taublast** [comment]
If one inspects the content of nugets the main difference between 2.80.2 (fails) and 1.68.3 (works) is that the first one is including native framewok with appropriate plist, the latter does not, explaining why it works.
Talking about the not working version while with the latest Xamarin.iOS it compiles, it fails to deploy with a "PackageInspectionFailed: Failed to load Info.plist from bundle". Searching web about it have found a very likely (multi-targeting - "Each probject has its own info.plist file. One of them was being copied into the bundle, causing the issue.") cause: (https://developer.apple.com/forums/thread/22330).
Now i don't know of it's skia sharp nuget project that has invalid settings (that became incompatible with newest xcode) of some new microsoft code is causing this incorrect behavior. Any thoughts about this direction?

**Comment by nowlights** [comment]
Hello,

same problem here with SkiaSharp on Windows...

**Comment by Rippletank** [comment]
Just to add another data point, this bug blocks deployment of Release builds for me on actual device with the PackageInspectionFailed error. But I don't get the error with a development key using a debug configuration build. I have separate development and deployment keys installed both pointing to the same app id. This is from both PC and using VS for mac.

There's been a wave of bugs I've had to wade through since starting on ios testing, was hopeful that another plist related bugfix ( [https://developercommunity.visualstudio.com/t/XamarinForms-ios:-IOException:-Sharing-/1355811](https://developercommunity.visualstudio.com/t/XamarinForms-ios:-IOException:-Sharing-/1355811) ) was going to fix it but still there's this issue with distribution builds.

Has anyone had any luck with preview releases of skiasharp? They seem to be quite regular.

**Comment by ahoke-cr** [comment]
> Has anyone had any luck with preview releases of skiasharp? They seem to be quite regular.

@Rippletank I ended up downgrading to SkiaSharp 1.68.0 in my iOS project while keeping the shared project and Android project at 2.80.2.

**Comment by Toine-db** [comment]
For my project I was able to fix it, and I'm still using SkiaSharp 2.80.2  (used in Svg.Skia)

I was also using SkiaSharp.Views, but managed to copy the single used method from that one to be able to remove the reference.
Not everything builds again.

Note; I'm using VS 16.9x  not 16.10 because I had to much issues with that version

Hope this helps

**Comment by Toine-db** [comment]
When I downgrade to Svg.Skia using SkiaSharp 1.68.x then it works fine on the emulator.
If I upgrade afterwards to 2.80.2 SKaiSharp the debug for emulators work fine as well.

But when I try to build 2.80.2 debug for a real iPhone everything goos wrong, from that point the emulator doesnt work any more....untill I downgrade and upgrade to 2.80.2 again... strange

So I can develop, but I can build releases anymore because the need to be build on iPhone device configuration... :-(

**Comment by nelutzu91** [comment]
@Toine-db and all others what had issues.... 
Downgrading to VS 16.9.2 worked for me. I was able to debug ios emulator, ios physical device. 
In my environment, I have VS professional 16.10, where I code... and VS enterprise 16.9.2 where I build iOS :D 
PS: if you want to spend some time, you can try downgrading VS one version at a time, maybe you find a newer version than 16.9.2 which works. Since I only tried 16.9.6 which didn't work and then I switched to 16.9.2, which worked for me before.

Hope this helps.

**Comment by Toine-db** [comment]
> @Toine-db and all others what have issues....
> Downgrading to VS 16.9.2 worked for me. I was able to debug ios emulator, ios physical device.
> In my environment, I have VS professional 16.10, where I code... and VS enterprise 16.9.2 where I build iOS :D
> PS: if you want to spend some time, you can try downgrading VS one version at a time, maybe you find a newer version than 16.9.2 which works. Since I only tried 16.9.6 which didn't work and then I switched to 16.9.2, which worked for me before.
> 
> Hope this helps.

What version of Skiasharp are you using?
Because I already downgraded to Vs 16.9.2

**Comment by dmsi-leo** [comment]
Latest VS 2019 (16.10.2) update seems to fix the issue. I updated SkiaSharp.Views.Forms  to 2.80.2 and was able build and deploy iOS project.

**Comment by nelutzu91** [comment]
> > @Toine-db and all others what have issues....
> > Downgrading to VS 16.9.2 worked for me. I was able to debug ios emulator, ios physical device.
> > In my environment, I have VS professional 16.10, where I code... and VS enterprise 16.9.2 where I build iOS :D
> > PS: if you want to spend some time, you can try downgrading VS one version at a time, maybe you find a newer version than 16.9.2 which works. Since I only tried 16.9.6 which didn't work and then I switched to 16.9.2, which worked for me before.
> > Hope this helps.
> 
> What version of Skiasharp are you using?
> Because I already downgraded to Vs 16.9.2, and debug on emulator only works but not on real devices.

SkiaSharp.Views.Forms 2.80.2 :)

**Comment by Toine-db** [comment]
> > > @Toine-db and all others what have issues....
> > > Downgrading to VS 16.9.2 worked for me. I was able to debug ios emulator, ios physical device.
> > > In my environment, I have VS professional 16.10, where I code... and VS enterprise 16.9.2 where I build iOS :D
> > > PS: if you want to spend some time, you can try downgrading VS one version at a time, maybe you find a newer version than 16.9.2 which works. Since I only tried 16.9.6 which didn't work and then I switched to 16.9.2, which worked for me before.
> > > Hope this helps.
> > 
> > 
> > What version of Skiasharp are you using?
> > Because I already downgraded to Vs 16.9.2, and debug on emulator only works but not on real devices.
> 
> SkiaSharp.Views.Forms 2.80.2 :)

Thanks, I just noticed I was using 16.9.6 instead of 16.9.2, seems this is what Im going todo today ;-)

**Comment by Toine-db** [comment]
@nelutzu91 You are a lifesaver, because of your comment I noticed the slight version difference and downgraded further from 16.9.6 to 16.9.2. and now everything works with Svg.Skia using SkiaSharp 2.80.2 

Emulators, real devices and even building releases and creating archives.

Thanks!

**Comment by reinhardsual** [comment]
> Latest VS 2019 (16.10.2) update seems to fix the issue. I updated SkiaSharp.Views.Forms to 2.80.2 and was able build and deploy iOS project.

It does fix the issues for me...

**Comment by Magendanz** [comment]
I get the following in my build log, but I believe it's the same issue.  Icon issues are resolved with 16.10.2, but this remains.

```
PackageInspectionFailed: Failed to load Info.plist from bundle at path /var/installd/Library/Caches/com.apple.mobile.installd.staging/temp.iyFP2M/extracted/TRF.iOS.app/Frameworks/libSkiaSharp.framework; Extra info about "/var/installd/Library/Caches/com.apple.mobile.installd.staging/temp.iyFP2M/extracted/TRF.iOS.app/Frameworks/libSkiaSharp.framework/Info.plist": Couldn't stat /var/installd/Library/Caches/com.apple.mobile.installd.staging/temp.iyFP2M/extracted/TRF.iOS.app/Frameworks/libSkiaSharp.framework/Info.plist: No such file or directory
```

If I try to upload to the App Store, I get this:

```
"altool" exited with code 1.
Invalid Bundle Structure - The binary file 'TRF.iOS.app/Frameworks/libSkiaSharp.framework/libSkiaSharp' is not permitted. Your app can’t contain standalone executables or libraries, other than a valid CFBundleExecutable of supported bundles. Refer to the Bundle Programming Guide at https://developer.apple.com/go/?id=bundle-structure for information on the iOS app bundle structure.
```

**Comment by Toine-db** [comment]
> I get the following in my build log, but I believe it's the same issue. Icon issues are resolved with VS 2019 16.10.2, but this remains.
> 
> ```
> PackageInspectionFailed: Failed to load Info.plist from bundle at path /var/installd/Library/Caches/com.apple.mobile.installd.staging/temp.iyFP2M/extracted/TRF.iOS.app/Frameworks/libSkiaSharp.framework; Extra info about "/var/installd/Library/Caches/com.apple.mobile.installd.staging/temp.iyFP2M/extracted/TRF.iOS.app/Frameworks/libSkiaSharp.framework/Info.plist": Couldn't stat /var/installd/Library/Caches/com.apple.mobile.installd.staging/temp.iyFP2M/extracted/TRF.iOS.app/Frameworks/libSkiaSharp.framework/Info.plist: No such file or directory
> ```
> 
> If I try to upload to the App Store, I get this:
> 
> ```
> "altool" exited with code 1.
> Invalid Bundle Structure - The binary file 'TRF.iOS.app/Frameworks/libSkiaSharp.framework/libSkiaSharp' is not permitted. Your app can’t contain standalone executables or libraries, other than a valid CFBundleExecutable of supported bundles. Refer to the Bundle Programming Guide at https://developer.apple.com/go/?id=bundle-structure for information on the iOS app bundle structure.
> ```

Try using VS 16.9.2

**Comment by ahoke-cr** [comment]
I updated to 16.9.10 but am now unable to deploy to either the simulator or a physical iPhone. I'm getting a package inspection error when deploying to a physical phone, and no specific error when deploying to the simulator. This is incredibly frustrating and I'm losing dev time trying to fix these issues that should not be issues.

**Comment by ahoke-cr** [comment]
> Try using VS 16.9.2

@Toine-db where did you find the installer for 16.9.2? I can only find it for 16.9.10.

**Comment by nelutzu91** [comment]
> > Try using VS 16.9.2
> 
> @Toine-db where did you find the installer for 16.9.2? I can only find it for 16.9.10.

@ahoke-cr , https://docs.microsoft.com/en-us/visualstudio/releases/2019/history#release-dates-and-build-numbers

**Comment by Magendanz** [comment]
> Try using VS 16.9.2

Note that there's no easy roll-back feature with VS.  You need to uninstall the current release and then do a clean install of the archived release, losing many of your configuration settings and cached credentials in the process.  Frankly, I'm not willing to risk destabilizing a dozen other projects that I maintain on this workstation just to work around this bug.  At this point I'm trying to scrounge up another build machine...or build in AppCenter.

**Comment by Toine-db** [comment]
> > Try using VS 16.9.2
> 
> Note that there's no easy roll-back feature with VS. You need to uninstall the current release and then do a clean install of the archived release, losing many of your configuration settings and cached credentials in the process. Frankly, I'm not willing to risk destabilizing a dozen other projects that I maintain on this workstation just to work around this bug. At this point I'm trying to scrounge up another build machine...or build in AppCenter.

You are absolutely right, but because of to much issues with Vs in the past I kept all my VS install very basic with only resharper. All other tools I use are not connected to VS anymore.

**Comment by dmsi-leo** [comment]
> @Magendanz
> > Try using VS 16.9.2
> 
> Note that there's no easy roll-back feature with VS. You need to uninstall the current release and then do a clean install of the archived release, losing many of your configuration settings and cached credentials in the process. Frankly, I'm not willing to risk destabilizing a dozen other projects that I maintain on this workstation just to work around this bug. At this point I'm trying to scrounge up another build machine...or build in AppCenter.

I am not sure if this can help but if your app is XF and not Xamarin.iOS try to remove unnecessary SkiaSharp references from all platform specific projects. In my case doing that and updating VS 2019 to the latest build along with updating all Nuget packages fixed all iOS deployment issues.

**Comment by Magendanz** [comment]
> I am not sure if this can help but if your app is XF and not Xamarin.iOS try to remove unnecessary SkiaSharp references from all platform specific projects. In my case doing that and updating VS 2019 to the latest build along with updating all Nuget packages fixed all iOS deployment issues.

I actually tried that yesterday, but the shared library in my XF project needs a Nuget package with a dependency on SkiaSharp and SkiaSharp.Views.Forms 2.80.2 or higher.  Even though SkiaSharp isn't explicitly installed or referenced in the iOS platform project, it still gets linked in.

**Comment by ahoke-cr** [comment]
Instead of downgrading from VS 16.10.2, I have maintained a reference to SkiaSharp 1.68.2 exclusively in my iOS-specific project. Android and the shared Xamarin project continue to reference SkiaSharp 2.80.2. I also had ARMv7s in my iOS Build under Supported Architectures which was causing issues with SkiaSharp, so my app is now only targeting ARMv7 + ARM64.

Maybe we can get a package update that actually fixes this issue for everyone instead of being forced to hack around a bad package release.

**Comment by ojuniour** [comment]
yep. I don't have SkiaSharp installed and for some reason it is still getting referenced in relation to the error.

**Comment by ahoke-cr** [comment]
@ojuniour make sure you are clearing build caches on both your Mac build host and your dev machine. After that, clean and rebuild the solution.

**Comment by ojuniour** [comment]
> @ojuniour make sure you are clearing build caches on both your Mac build host and your dev machine. After that, clean and rebuild the solution.

What's the most effective way of clearing out cache on the Mac if I'm connecting my build through VS on windows.

**Comment by ahoke-cr** [comment]
@ojuniour I use Google Remote Desktop to remotely connect to my Mac build host. You can also use SSH to execute commands on your Mac. More details [here](https://docs.microsoft.com/en-us/xamarin/ios/get-started/installation/windows/connecting-to-mac/troubleshooting#clearing-the-broker-idb-build-and-designer-agents-on-the-mac)

**Comment by ojuniour** [comment]
> @ojuniour I use Google Remote Desktop to remotely connect to my Mac build host. You can also use SSH to execute commands on your Mac. More details [here](https://docs.microsoft.com/en-us/xamarin/ios/get-started/installation/windows/connecting-to-mac/troubleshooting#clearing-the-broker-idb-build-and-designer-agents-on-the-mac)

Clearing both caches don't work either.

**Comment by ahoke-cr** [comment]
> Clearing both caches don't work either.

@ojuniour I would manually check your csproj files for any references, as well as verifying that you don't have any other packages that also reference SkiaSharp. Make sure you're clearing your `bin` and `obj` folders in your Xamarin iOS platform project. You can also go to AppData folders and remove the package caches to ensure that there's no way the SkiaSharp.dll could be referenced during build.

**Comment by ojuniour** [comment]
> > Clearing both caches don't work either.
> 
> @ojuniour I would manually check your csproj files for any references, as well as verifying that you don't have any other packages that also reference SkiaSharp. Make sure you're clearing your `bin` and `obj` folders in your Xamarin iOS platform project. You can also go to AppData folders and remove the package caches to ensure that there's no way the SkiaSharp.dll could be referenced during build.

Already did all that ... except checking the csproj for reference. I'll thoroughly look into that.
Thanks.

**Comment by dcon0610** [comment]
For me, the issue was that I had a fresh install of visual studio in windows (new pc) and only had the skia sharp 2.80.2 installed. Tried everything else, what solved it was rolling back to skia sharp 1.68.0 and it worked. Then, rolling back forward to 2.80.2 still works. Hopefully an update will solve others wasting approx 8 hours of dev time.

**Comment by Magendanz** [comment]
Today's v16.10.3 release of Visual Studio 2019 (with Xamarin.iOS 14.20.24) seems to have fixed this issue for me.  Can anyone else confirm?

**Comment by skorpsim** [comment]
The v16.10.3 release fixed the issue for me as well. Cleared the Mac's build cache and on Windows the bin and obj, folders of the iOS project to make sure that clean rebuilds work as expected.

**Comment by taublast** [comment]
Can confirm this has been fixed in VS 16.10.3 + Xamarin.iOS 14.20.0.2.4. 
Still using XCode 12.5 (latest is 12.5.1).
Surprisingly fixed in not mentioned in VS release notes.

**Comment by bondarenkod** [comment]
https://developercommunity.visualstudio.com/t/Xamarin-Forms-iOS-SkiaSharp-DirectoryNot/1424416 
it seems like fixed now. Working like a charm. Finally. Just tested.
![msedge_30_06_2021__10_51_07__9bfb4f84-7743-417b-8f70-82e0c4d0402e](https://user-images.githubusercontent.com/3184414/123922820-16be7d00-d991-11eb-9d55-c3ca71355cb0.png)
PS: I have started testing all VS updates more than one year ago. Once I updated the VS and found a bug a week after that, so I had to reinstall all the VS installations on my working environment. I spent more than one working day (by hours) to recover my working environment.

**Comment by cbs-cbt** [comment]
I can confirm the issue has been fixed in VS 16.10.3

**Comment by ojuniour** [comment]
I am also confirming that it had been fixed in 16.10.3 also.

**Comment by mattleibow** [comment]
This is good news! I'll close this issue for now. If this happens again, please open an new issue. It might help as well if suddenly a build stops working after updating VS and works again when rolling back to either open an issue in:
 - https://github.com/xamarin/xamarin-android
 - https://github.com/xamarin/xamarin-macios
